pkg/syncer/cnsoperator/controller/cnsnodevmattachment/cnsnodevmattachment_controller.go

```diff
@@ -382,9 +389,16 @@ func (r *ReconcileCnsNodeVMAttachment) Reconcile(ctx context.Context,
 
 			if attachErr != nil {
 				recordEvent(ctx, r, instance, v1.EventTypeWarning, "")
-				return reconcile.Result{RequeueAfter: timeout}, "NotComputed", nil
+				return reconcile.Result{RequeueAfter: timeout}, faulttype, nil
 			}
 
+			err = addFinalizerToPVC(ctx, r.client, pvc)
+			if err != nil {
+				msg := fmt.Sprintf("failed to add %q finalizer on the PVC with volumename: %q on namespace: %q. Err: %+v",
+					cnsoperatortypes.CNSPvcFinalizer, instance.Spec.VolumeName, instance.Namespace, err)
+				recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+				return reconcile.Result{RequeueAfter: timeout}, csifault.CSIInternalFault, nil
```

Comment:
Say the workflow failed to add the finalizer to the PVC, can you explain how it will be added back again on the next reconcile invocation?

Reply:
Looks like this part of the code was not getting invoked again since a successful attachvolume operation would mark the Status.Error = "" & Status.Attached=true. After taking another look, I have moved  this piece of code early in the workflow. Only after a successful finalizer update on the PVC, attachVolume will be called. 

With this change, if addFinalizerToPVC fails, then the instance will be marked with Error=err(addFinalizerToPVCError or similar) and the instance.Status.Attached != true. The next reconcile will invoke this code path again, if addFinalizer  succeeds, AttachVolume will be called and instance will be marked Attached and Error="".